datafusion/row/src/writer.rs

```diff
@@ -349,7 +349,7 @@ pub(crate) fn write_field_utf8(
     let from = from.as_any().downcast_ref::<StringArray>().unwrap();
     let s = from.value(row_idx);
     let new_width = to.current_width() + s.as_bytes().len();
```

Comment:
Here, we can use `let new_width = to.current_width() + s.as_bytes().len() + 8`

Reply:
And in the `set_utf8` or `set_binary`, we can use 

```
        self.varlena_offset += (size + 8);
        self.varlena_width += (size +8);
```
to update the offset and width.

Reply:
> And in the `set_utf8` or `set_binary`, we can use
> 
> ```
>         self.varlena_offset += (size + 8);
>         self.varlena_width += (size +8);
> ```
> 
> to update the offset and width.

That was also my vision. But my concern here if we add 8 extra bytes per field, it may bloat the row size significantly?

Reply:
@comphead you can go through the code of `set_utf8`.
```
    fn set_utf8(&mut self, idx: usize, value: &str) {
        self.assert_index_valid(idx);
        let bytes = value.as_bytes();
        let size = bytes.len();
        // this line: write the offset and size to the row_buffer
        self.set_offset_size(idx, size as u32);
        let varlena_offset = self.varlena_offset;
        self.data[varlena_offset..varlena_offset + size].copy_from_slice(bytes);
        self.varlena_offset += size;
        self.varlena_width += size;
    }
```
Not per field, and variable length data type like utf8 and binary.
From the @yjshen  codebase, I think the layout of utf8 and binary should be
```
─ ─ ─ ─ ─ ─  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
│ offset and size│     utf8 data or binary data │
─ ─ ─ ─ ─ ─  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
```

Reply:
I figure out the corner case to panic for your fix.
You can add the test with string data type, but the each value of string empty, just like that
```
arrow::array::StringArray::from(vec!["","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","",""]);
```

Reply:
@liukun4515 can elaborate on why there are two '+ 8' suggested?

```
    self.varlena_offset += (size + 8);
    self.varlena_width += (size +8);
```

Reply:
Please ignore my comment, I have a error thought about the row layout when writing the comments